File: ttt.js
var output = document.getElementById("output");
function print(s) {
  output.innerText += s;
}

var board = ['_', '_', '_',
             '_', '_', '_',
             '_', '_', '_'];

var current_player = 'X';

// NOTE(dbp 2015-12-15): Takes index, tells you whether
// it is at the end of a line.
function end_of_line(idx) {
  if (typeof idx === "undefined") {
    print("usage: end_of_line(idx)\n");
  }

  return idx % 3 === 2;
}

function print_board(board) {
  if (typeof board === "undefined") {
    print("usage: print_board(some_board)\n");
  }
  board.map(function (x, idx) {
    print(x + " ");
    if (end_of_line(idx)) {
      print("\n");
    }
  });

  print("\n");
}

function swap_player(player) {
  if (typeof player === "undefined") {
    print("usage: swap_player(some_board)\n");
  }
  if (player === 'X') {
    return 'O';
  } else {
    return 'X';
  }
}

// NOTE(dbp 2015-12-15): position is 0-indexed on board. valid values are 0 - 8.
function legal_move(board, position) {
  if (typeof board === "undefined" || typeof position === "undefined") {
    print("usage: legal_move(some_board, a_position)\n");
  }
  if (position < 0 || position > 8) {
    print("position " + position + " is not on board!\n");
    return false;
  }
  if (board[position] !== '_') {
    return false;
  } else {
    return true;
  }
}

function start_game() {
  board = ['_', '_', '_',
           '_', '_', '_',
           '_', '_', '_'];
}

function human_turn(position) {
  if (typeof position === "undefined") {
    print("usage: my_turn(a_position)\n");
  }
  var current_player = 'X';
  if (legal_move(board, position)) {
    board[position] = current_player;
    print_board(board);
    current_player = swap_player(current_player);
  } else {
    print("Position is not legal. Please try another.\n");
    print_board(board);
  }

}

function auto_turn() {
  var open_spaces = board.map(function(_,idx) {
    return idx;
  }).filter(function(idx) {
    // NOTE(dbp 2015-12-15): Unoccupied spots.
    return board[idx] === '_';
  });

  var random_position = Math.floor(Math.random() * open_spaces.length);

  if (legal_move(board, random_position)) {
    board[random_position] = current_player;
    print_board(board);
    current_player = swap_player(current_player);
  } else {
    their_turn();
  }
}

function game_is_over(board) {
  if (typeof board === "undefined") {
    print("usage: game_is_over(some_board)\n");
  }
  // TODO(dbp 2015-12-15): Write this!
  return false;
}

File: ttt.py
from random import randrange
import sys

prnt = sys.stdout.write

board = ['_', '_', '_',
         '_', '_', '_',
         '_', '_', '_']

def end_of_line(idx):
    return idx % 3 == 2

def print_board(board):
  """ This function prints out the board """

  def print_one(pair):
      (idx, x) = pair
      prnt(x + " ")
      if end_of_line(idx):
          prnt("\n")
  map(print_one, enumerate(board))

  # for idx in range(len(board)):
  #     prnt(board[idx] + " ")
  #     if (idx + 1) % 3 == 0:
  #         prnt("\n")

  prnt("\n")


def swap_player(player):
  if player == 'X':
    return 'O'
  elif player == 'O':
    return 'X'
  else:
    raise ValueError("swap_player expected 'X' or 'O'")

# TODO(dbp 2015-12-15): position is 0-indexed on board. valid values are 0 - 8.
def legal_move(board, position):
  if position < 0 or position > 8:
    prnt("position " + str(position) + " is not on board!\n");
    return False
  if board[position] != '_':
    return False
  else:
    return True

def start_game():
  board = ['_', '_', '_',
           '_', '_', '_',
           '_', '_', '_']

def human_turn(position, current_player):
  if legal_move(board, position):
    board[position] = current_player
    print_board(board)
    return swap_player(current_player)
  else:
    prnt("Position is not legal. Please try another.\n")
    print_board(board)

def auto_turn(current_player):
  open_spaces = filter(lambda idx: board[idx] == '_', range(len(board)))

  # TODO(dbp 2015-12-15): only choose from open spaces
  random_position = randrange(len(open_spaces))

  if legal_move(board, random_position):
    board[random_position] = current_player
    print_board(board)
    return swap_player(current_player)
  else:
    auto_turn();

def game_is_over(board):
  # TODO(dbp 2015-12-15): Write this!
  return False
